### ISSUE TITLE
Discussion on some key questions with a DL framework

### ISSUE THREAD

**Issue by wangkuiyi** [PaddlePaddle/Paddle#2454]
This issue comes from https://github.com/PaddlePaddle/Paddle/pull/2445#issuecomment-308007213

I think it would be easier to subdivide this document into the following topics, and try to figure out an optimal design for each topic:

1. memory management
   1. class Place, inspired by Majel
   1. Allocation and Allocators
   1. unified malloc and allocation API for GPU and CPU --
       ```cpp
       p=malloc(Place pl, ...);
       used(pl);
       free(p); 
       ```
1. Tensor
   1. consider re-use existing ones, including mshadow, Eigen, porting Majel, or write a wrapper of these libraries.
1. Expression Template
   1. Expression Template is important for performance optimization.
   1. Is there any (what is the) difference between above libraries -- mshadow and Eigen. (Majel doesn't complete implementation of Expression Template.)
   1. Which is better?
1. Ops and Variables
   1. TensorFlow's Ops take Tensor, a wrapper of Eigen, as its inputs and outputs.
   1. Caffe2 and PyTorch take *Variable* as Ops' inputs and outputs.
   1. A Variable includes a tensor for the forward algorithm and another gradient tensor for the backward algorithm.  The difference seems that TensorFlow is a general-purpose framework, whereas Caffe2 and PyTorch are DL-specific.
   1. Which approach should PaddlePaddle take?
1. Ops and Gradient Ops
   1. In TensorFlow, an Op is a general concept -- all computations are represented by ops.
   1. In Caffe2, each Op has one or more corresponding GradientOps.
   1. Which approach should PaddlePaddle take?
1. Ops and Kernels
   1. TensorFlow separate Op's signature (as OpDef) and its implementations (as OpKernel)
   1. others might not have such clear separation.
   1. Which approach should PaddlePaddle take.
1. Execution engine
   1. How TensorFlow and other solutions parse the network definition and create a network?
   1. How TensorFlow and other solutions execute the training algorithm over the network while creating/managing the memory of Variables?

**Comment by jacquesqiao** [comment]
5，7

**Comment by hedaoyuan** [comment]
3
